docs/rfc/rfc-105-event-observer.md

```diff
@@ -0,0 +1,240 @@
+which seems to signal that its functionality should be decentralized.
+
+### Strongly Typed Event Observer
+
+A simple alternative pattern here would be to define an **event observer**
+interface for every major component of the system capable of producing events.
+For consensus state, this may end up looking something like:
+
+```golang
+package consensus
+
+// StateObserver is specific to the consensus.State struct, and all of its
+// methods are called from within consensus.State instead of using an "event
+// bus". This allows for greater compile-time guarantees through doing away with
+// the generic pub/sub mechanism in the event bus.
+//
+// Note how all methods are infallible (i.e. they do not return any errors).
+// This is functionally equivalent to the fire-and-forget pattern implemented by
+// the event bus.
+//
+// Also note how method names are prefixed by the name of the relevant producer
+// of events (in this case "ConsensusState", corresponding to the
+// consensus.State struct). This is intentional to allow composition of
+// observers of multiple different components without function names clashing.
+//
+// Finally, given that this is just straightforward Go, it is up to either the
+// caller or the callee to decide how to handle the concurrency of certain
+// events. The event bus approach, by contrast, is always concurrent and relies
+// on Go channels, which could end up filling up and causing back-pressure into
+// the caller (already observed in slow WebSocket subscribers).
+type StateObserver interface {
+    ConsensusStateNewRoundStep(ev EventDataRoundState)
+    ConsensusStateTimeoutPropose(ev EventDataRoundState)
+    ConsensusStateTimeoutWait(ev EventDataRoundState)
+    // ...
+}
+```
+
+And on `consensus.State` one could easily either supply such an observer in the
+constructor, or define a new method that allows one to set the observer, e.g.:
+
+```golang
+package consensus
+
+func (cs *State) SetObserver(obs StateObserver) {
+    cs.observer = obs
+}
+```
+
+Then, instead of publishing events via the event bus, one would simply do the
+following:
+
+```diff
+-        if err := cs.eventBus.PublishEventNewRoundStep(rs); err != nil {
+-            cs.Logger.Error("failed publishing new round step", "err", err)
+-        }
++        // Notify the observer
++        cs.observer.ConsensusStateNewRoundStep(rs)
+```
+
+### Comparing "Subscription" Interfaces
+
+The `EventBus` offers two slightly different subscription mechanisms for events,
+both of which rely on Go channels under the hood for their implementation.
+
+- [`Subscribe`]
+- [`SubscribeUnbuffered`]
+
+The observer pattern proposed in this RFC can facilitate both patterns and more,
+given that one has the option of intervening synchronously in a blocking way
+when the "publisher" calls the observer event handler. Using the proposed
+observer pattern, it is up to either the publisher or the observer to implement
+their own concurrency.
+
```

Comment:
Another great benefit of this approach with respect to a go-channel-based one is the following.
In the near future, we may want to "standardise" reactor boundaries to improve the modularity of the system... allowing e.g., reactors in different processes, or even different programming languages for some reactors. The fact that the reactors interact via go channels, would complicate that effort, as go channels are particular to go and hard to handle in e.g., Protobufs, FFI, etc.

I think we can also include this argument in favor of this approach.